lighthouse-core/computed/layout-shift-variants.js

```diff
@@ -0,0 +1,158 @@
+ * promoted to a Core Web Vital.
+ * @see https://web.dev/better-layout-shift-metric/
+ * @see https://github.com/mmocny/web-vitals/wiki/Snippets-for-LSN-using-PerformanceObserver
+ */
+
+class LayoutShiftVariants {
+  /**
+   * Calculates the average LayoutShift score across clusters of LayoutShift
+   * events, where a new cluster is created when there's a gap of at least 5s
+   * since the last LayoutShift event.
+   * @param {Array<LayoutShiftEvent>} layoutShiftEvents
+   * @return {number}
+   */
+  static avgSessionGap5s(layoutShiftEvents) {
+    if (layoutShiftEvents.length === 0) return 0;
+
+    let score = 0;
+    let clusterCount = 0;
+    let prevTs = Number.NEGATIVE_INFINITY;
+
+    for (const event of layoutShiftEvents) {
+      if (event.ts - prevTs > 5_000_000) clusterCount++;
+      prevTs = event.ts;
+      score += event.score;
+    }
+
+    return score / clusterCount;
+  }
+
+  /**
+   * Calculates cumulative layout shifts per cluster (session) of LayoutShift
+   * events -- where a new cluster is created when there's a gap of at least
+   * `gapMs` ms since the last LayoutShift event or the cluster has reached
+   * `limitMs` ms long -- and returns the max LayoutShift score found.
+   * `limitMs` is optional if no limit is needed.
+   * @param {Array<LayoutShiftEvent>} layoutShiftEvents
+   * @param {number} gapMs
+   * @param {number} [limitMs]
+   */
+  static maxSession(layoutShiftEvents, gapMs, limitMs = Number.POSITIVE_INFINITY) {
+    let maxScore = 0;
+    let currentClusterScore = 0;
+    let firstTs = Number.NEGATIVE_INFINITY;
+    let prevTs = Number.NEGATIVE_INFINITY;
+
+    for (const event of layoutShiftEvents) {
+      if (event.ts - firstTs > limitMs * 1000 || event.ts - prevTs > gapMs * 1000) {
+        firstTs = event.ts;
+        currentClusterScore = 0;
+      }
+      prevTs = event.ts;
+      currentClusterScore += event.score;
+      maxScore = Math.max(maxScore, currentClusterScore);
+    }
+
+    return maxScore;
+  }
+
+  /**
+   * Returns the maximum cumulative layout shift in any `windowMs` ms window in
+   * the trace.
+   * @param {Array<LayoutShiftEvent>} layoutShiftEvents
+   * @param {number} windowMs
+   * @return {number}
+   */
+  static maxSliding(layoutShiftEvents, windowMs) {
+    let maxScore = 0;
+    let currentWindowScore = 0;
+    const windowEvents = [];
+
+    for (const event of layoutShiftEvents) {
+      while (windowEvents.length && event.ts - windowEvents[0].ts > windowMs * 1000) {
+        currentWindowScore -= windowEvents[0].score;
+        windowEvents.shift();
+      }
+      windowEvents.push(event);
+      currentWindowScore += event.score;
+      maxScore = Math.max(maxScore, currentWindowScore);
+    }
+
+    return maxScore;
+  }
+
+  /**
+   * Returns all main process LayoutShift events.
+   * @param {LH.Artifacts.TraceOfTab} traceOfTab
+   * @return {Array<LayoutShiftEvent>}
+   */
+  static getLayoutShiftEvents(traceOfTab) {
```

Comment:
cc @adamraine maybe there's interplay between your planned refactor and this PR? seems like CLS itself could be a "layout shift variant" `maxSession(allEvents, Infinity, Infinity)`

heck our current impl should probably use this :)

Reply:
My plan for the refractor was to add the layout shift events for all frames to `TraceOfTab` and then filter/aggregate the events in CLS and CLS-AF. The `had_recent_input` logic would be in `TraceProcessor`.

I could use this `getLayoutShiftEvents` instead if it wasn't specific to main frame events. Maybe pick out the main frame events in `_compute`?  It would also make sense to move this function elsewhere since it wouldn't be specific to this file.

Either way, we should probably be refactoring main frame CLS so we don't repeat the `had_recent_input` logic.

Reply:
> I could use this `getLayoutShiftEvents` instead if it wasn't specific to main frame events. Maybe pick out the main frame events in `_compute`?

yeah, that would work fine

> My plan for the refractor was to add the layout shift events for all frames to `TraceOfTab` and then filter/aggregate the events in CLS and CLS-AF. The `had_recent_input` logic would be in `TraceProcessor`.

@patrickhulce are we just making our job harder for FR, though? I wonder if any correction we do should be pushed back even further...like should the correction be done during gathering, when everything knows whether or not it was a `navigation` and so can confidently correct `had_recent_input`?

OTOH for `timeOrigin` stuff, TraceProcessor is going to have to be given some level of knowledge about what kind of run the trace comes from anyways...

Reply:
> are we just making our job harder for FR, though? I wonder if any correction we do should be pushed back even further...like should the correction be done during gathering, when everything knows whether or not it was a navigation and so can confidently correct had_recent_input?

When we get there, the workaround will need to be an option for sure. Unfortunately, I think we're just kinda screwed for FR no matter what we do. I called this out in @adamraine's PR too that if we're auto enabling emulation with timespans we'll be stuck in between both of these (we want to ignore had_recent_input for the emulation change just like navigation but the user could have had real input too :/) 

Either way, I'm fine with whatever is most comfortable for these two immediate PRs because FR trace processing is a massive project I don't expect to be significantly affected by this decision here.

> edit: or we just share a getLayoutShiftEvents function between the various CLS impls and don't try to correct the trace at all

This was roughly what I had in mind originally but with `options: {ignoreInitialHadRecentInputSequence: boolean}` or something

Reply:
Was this settled? It seems like we're still just duplicating at the moment, and I'm a little unclear of how it will be resolved in a followup.

Reply:
> Was this settled? It seems like we're still just duplicating at the moment, and I'm a little unclear of how it will be resolved in a followup.

this is duplicating but based on https://github.com/GoogleChrome/lighthouse/pull/12034#issuecomment-771889207 (and the conversation before it) it sounded like we still need a third copy and that'll be the time to do the refactor for them to all share some level of implementation. cc @adamraine

Reply:
I can also do the follow up refactor if there's too much else going on right now.

Reply:
> it sounded like we still need a third copy and that'll be the time to do the refactor for them to all share some level of implementation

Yeah, I think we should settle the duplication in a follow-up.

Reply:
> I can also do the follow up refactor if there's too much else going on right now.

I started working on this locally, but I was waiting for this to land so I could use `getLayoutShiftEvents`.